### PR TITLE
add docker volume size to node definition

### PR DESCRIPTION
**Pull request by teemow**
this makes the size of the docker volume that is attached to each node.
docker uses this volume to fetch images and depending on the size of the
images being used in the cluster the users would like to configure the
size of this volume.

disclaimer: the data structure of nodes/workers is messed up already.
there are differences between aws/azure and kvm whether the nodes can be
configured individually or not. this is a mess and this addition will
not aim to fix this. this will be changed with the work on node pools.

### CODE REVIEW

File: definitions.yaml [hunk 1]
```diff
@@ -55,6 +55,19 @@ definitions:
               default:
                 description: The instance type used in new cluster, if not specified
                 type: string
+          docker_volume_size:
+            description: Size of the docker volume attached to the node. This volume is used to store docker images.
```
Comment (MarcelMue) [inline definitions.yaml]: The line wrapping is enforced here, so these lines are probably too long.

Comment (teemow) [reply definitions.yaml]: there is only a warning about the length. the errors were about trailing whitespace. I've removed it.

**Comment by teemow** [comment]
@oponder https://github.com/giantswarm/giantswarm/issues/4105#issuecomment-418308852

File: details/CLUSTER_DEFINITION.md [hunk 2]
```diff
@@ -62,15 +64,17 @@ This example here shows a cluster definition in JSON format, as it may be submit
...
+                "docker_volume_size": {"size_gb": 100},
             },
             "labels": {
                 "nickname": "first node"
             }
         },
         {
             "aws": {
-              "instance_type": "m3.large"
+                "instance_type": "m3.large"
```
Comment (oponder) [inline details/CLUSTER_DEFINITION.md]: missing a comma at the end of the line here.

File: details/CLUSTER_DEFINITION.md [hunk 3]
```diff
@@ -62,15 +64,17 @@ This example here shows a cluster definition in JSON format, as it may be submit
     "workers": [
         {
             "aws": {
-              "instance_type": "m3.large"
+                "instance_type": "m3.large"
```
Comment (oponder) [inline details/CLUSTER_DEFINITION.md]: missing a comma at the end of the line here.

File: details/CLUSTER_DEFINITION.md (hunk 2 above)
Comment (teemow) [reply details/CLUSTER_DEFINITION.md]: actually I've put the `docker_volume_size` into the provider specific configuration and thats wrong. fixed it.

File: details/CLUSTER_DEFINITION.md (hunk 3 above)
Comment (teemow) [reply details/CLUSTER_DEFINITION.md]: same here

Comment (teemow) [reply details/CLUSTER_DEFINITION.md]: talking about: https://github.com/giantswarm/api-spec/pull/108/files#r214950270

File: definitions.yaml [hunk 4]
```diff
@@ -55,6 +55,19 @@ definitions:
               default:
                 description: The instance type used in new cluster, if not specified
                 type: string
+          docker_volume_size:
```
Comment (marians) [inline definitions.yaml]: I recommend highly to rename `docker_volume_size` to `docker_volume_size_gb`, in line with the `size_gb` attribute we have in memory and storage info on nodes. The `_gb` eliminates the guesswork regarding what unit is used. Assuming that we are using Gigabyte here.

**Comment by teemow** [comment]
I like that very much and changed it. PTAL

I've also aligned the structure in the info endpoint as well.

**Comment by teemow** [comment]
Please also take a look at: https://github.com/giantswarm/api-spec/pull/108/commits/3c03013639666e5476c18f03c61dff4149571456

We've deprecated the storage object of workers in favor of configuring the volumes individually.

File: definitions.yaml [hunk 5]
```diff
@@ -55,6 +55,27 @@ definitions:
               default:
                 description: The instance type used in new cluster, if not specified
                 type: string
+          volumes:
+            description: Information related to volumes on worker nodes
+            type: object
+            properties:
+              docker:
+                description: Information about the docker volume attached to the worker node. This volume is used to store docker images.
```
Comment (marians) [inline definitions.yaml]: > This volume is used to store docker images.

Is this correct/complete? I reckon that running container's overlay file systems are also placed there.

File: definitions.yaml [hunk 6]
```diff
@@ -196,7 +217,26 @@ definitions:
...
+              option is deprecated it will be used as the docker volume size
+              but only if the docker volume size isn't defined itself.
+      volumes:
+        description: Configuration for volumes that are attached to worker nodes.
+        type: object
+        properties:
+          docker:
+            description: |
+              Volume for the docker daemon. Usually mounted at `/var/lib/docker`
+              The docker images are stored in this volume.
```
Comment (marians) [inline definitions.yaml]: Please add a period after `/var/lib/docker`

File: details/CLUSTER_DEFINITION.md [hunk 7]
```diff
@@ -136,6 +150,7 @@ After creation using the definition above, the code below shows the completed cl
...
 - `workers[n].memory`: Memory definition object. This object can currently contain only one attribute `size_gb`, which indicates the RAM size in Gigabytes as an integer. For clusters running on AWS, this attribute is ignored on cluster creation.
-- `workers[n].storage`: Storage definition object. Storage here refers to the local disk storage of a worker node. The definition object can currently contain only one attribute `size_gb`, which indicates the local storage size in Gigabytes as an integer. For clusters running on AWS, this attribute is ignored on cluster creation.
+- `workers[n].volumes`: Volumes definition object. The volumes are ephemeral storage that is attached to each worker node. 
+- `workers[n].volumes[n].docker`: Docker volume size definition object. This refers to the size of the disk attached to the node where docker images are stored. This volume is usually attached to `/var/lib/docker` and the docker daemon uses this location for storing images. The definition object can currently contain only one attribute `size_gb`, which indicates the docker volume size in Gigabytes as an integer.
```
Comment (marians) [inline details/CLUSTER_DEFINITION.md]: `volumes[n].docker` indicates that `volumes` is an array, which it isn't. Please change to `volumes.docker`.

Comment (marians) [reply details/CLUSTER_DEFINITION.md]: > ... for storing images.

See above comment regarding usage of this volume

Comment (teemow) [reply details/CLUSTER_DEFINITION.md]: resolved

File: definitions.yaml (hunk 6 above)
Comment (teemow) [reply definitions.yaml]: resolved

File: definitions.yaml (hunk 5 above)
Comment (teemow) [reply definitions.yaml]: resolved

**Comment by marians** [comment]
@teemow Once you get back to this, would you like to mark some of the new attributes as `required`?

Please also update the main description of this PR so that it will explain after merging what has been changed. Ideally it would contain a list of endpoints that have been modified.

**Comment by oponder** [comment]
@teemow can I close this? I think it's not gonna be implemented any time soon right? ;)

**Pull request closed without merging**